components/Layout.tsx

```diff
@@ -0,0 +1,239 @@
+import { ReactNode, useEffect, useState } from 'react';
+import Link from 'next/link';
+import { useRouter } from 'next/router';
+import { ApolloProvider, useApolloClient } from '@apollo/client/react';
+import { ApolloClient, InMemoryCache } from '@apollo/client';
+import styled from '@emotion/styled';
+import {
+  DirectionalHint,
+  Dropdown,
+  IDropdownStyles,
+  IDropdownOption,
+  IRenderFunction,
+  Icon,
+  NeutralColors,
+  SearchBox,
+  TooltipDelay,
+  TooltipHost,
+} from '@fluentui/react';
+
+import {
+  getEndpointFromQuery,
+  GraphQLEndPoint,
+  GRAPHQL_ENDPOINTS,
+} from 'lib/graphQLEndPoint';
+import {
+  BlockByIndexQuery,
+  BlockByIndexDocument,
+  BlockByHashQuery,
+  BlockByHashDocument,
+  TransactionByIdQuery,
+  TransactionByIdDocument,
+} from 'src/gql/graphql';
+
+import Wrapper from 'components/Wrapper';
+import { CommonPageProps } from 'lib/staticGeneration';
+import useIdFromQuery from 'lib/useIdFromQuery';
+
+export default function Layout({
+  children,
+  staticEndpoint,
+}: CommonPageProps & {
+  children: ReactNode;
+}) {
+  const { isReady, query } = useRouter();
+  const [endpoint, setEndpoint] = useState(staticEndpoint);
+  const [client, setClient] = useState<ApolloClient<object> | null>(null);
+  useEffect(() => {
+    if (!endpoint && isReady) {
+      setEndpoint(getEndpointFromQuery(query));
+    }
+  }, [endpoint, isReady, query]);
+  useEffect(() => {
+    if (endpoint) {
+      setClient(
+        new ApolloClient({
+          cache: new InMemoryCache(),
+          uri: endpoint?.uri,
+        })
+      );
+    }
+  }, [endpoint]);
+  if (endpoint) {
+    if (!client) return null;
+
+    return (
+      <ApolloProvider client={client}>
+        <LayoutContainer>
+          <NavBar endpoint={endpoint} />
+          <Wrapper>{children}</Wrapper>
+        </LayoutContainer>
+      </ApolloProvider>
+    );
+  } else {
+    // redirect to the first endpoint
+    return <LayoutContainer>{children}</LayoutContainer>;
+  }
+}
+
+const dropdownStyles: Partial<IDropdownStyles> = {
+  dropdown: { width: 120 },
+};
+
+export function NavBar({ endpoint }: { endpoint: GraphQLEndPoint }) {
+  const client = useApolloClient();
+  const { push, query } = useRouter();
+  const [searchBoxValue, setSearchBoxValue] = useState('');
+  const queriedId = useIdFromQuery(query);
+  useEffect(() => {
+    setSearchBoxValue(queriedId ?? '');
+  }, [queriedId]);
+  const onSearch = async (value: string) => {
+    value = value.trim();
+    if (value.match(/^[0-9a-fA-F]{64}$/)) {
+      const data = await client
+        .query<BlockByHashQuery>({
+          query: BlockByHashDocument,
+          variables: { hash: value },
+        })
+        .catch(() => null);
```

Comment:
Is it okay if we ignore that exception here?

Reply:
I thought we didn't need error handling here since if there's an network error it would be evident on the view and if there is a graphql error it would be found during testing, but now that I think about it a simple check might help, so I added a commit with the error handling.